### PR TITLE
Rework the ScriptEngine shutdown process, Fix crash in AC Scripts

### PR DESCRIPTION
**Pull request by ZappoMan**
* Fixes a crash on stop/reload script in interface
* Fixes a crash on shutdown in interface
* Fixes a crash in startup of AC Scripts

### CODE REVIEW

File: libraries/script-engine/src/ScriptEngine.cpp [hunk 1]
```diff
@@ -117,20 +118,28 @@ ScriptEngine::~ScriptEngine() {
     }
 }
 
+class MyWorkerThread : public QThread {
+public:
+    MyWorkerThread(QObject* parent = nullptr) : QThread(parent) { qDebug() << "MyWorkerThread::MyWorkerThread() this:" << this; }
+    ~MyWorkerThread() { qDebug() << "MyWorkerThread::~MyWorkerThread() this:" << this; }
+};
```
Comment (birarda) [inline libraries/script-engine/src/ScriptEngine.cpp]: Assuming you only need the `MyWorkerThread` class to verify the QThread is doing what you expect?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6521de81634ac390fd8dcc5a83806bb193d2a6b1 available at: https://gist.github.com/bf23f5066b9358fbe272<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5811.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5811.dmg

File: libraries/script-engine/src/ScriptEngine.cpp (hunk 1 above)
Comment (ZappoMan) [inline libraries/script-engine/src/ScriptEngine.cpp]: correct - I plan to remove this. this is why this is a DO NOT MERGE

**Comment by hifi-gustavo** [comment]
Build console output for commit: bdae3e420b21e80dc71c9ebe121b2dbe6717d9c0 available at: https://gist.github.com/bf23f5066b9358fbe272<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5811.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5811.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 41caa36038084d53d5be0b28b572c9b2c73307c6 available at: https://gist.github.com/bf23f5066b9358fbe272<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5811.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5811.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 016a5e5f0a0607ead6a7c4ada9be4c2f600a2240 available at: https://gist.github.com/bf23f5066b9358fbe272<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5811.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5811.dmg

**Comment by howard-stearns** [comment]
I tried to smoke test the mac build, but on exit I always crash as follows. While this is presumably not related, I don't know if that crash is masking the ones that we're looking for.

Dyld Error Message:
  can't resolve symbol _CleanupConnexionHandlers in /Volumes/Interface PR 5811 - Mac Build: 1512/interface.app/Contents/MacOS/interface because dependent dylib #1 could not be loaded

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   dyld                          	0x00007fff60014075 dyld_fatal_error + 1
1   dyld                          	0x00007fff600172a8 dyld::fastBindLazySymbol(ImageLoader**, unsigned long) + 180
2   libdyld.dylib                 	0x00007fff8dbe23ba dyld_stub_binder + 282
3   ???                           	0x000000010e1ae008 0 + 4531609608
4   io.highfidelity.Interface     	0x000000010d72ce98 Application::~Application() + 1336
5   io.highfidelity.Interface     	0x000000010d82a2bd main + 733
6   libdyld.dylib                 	0x00007fff8dbe45c9 start + 1

**Comment by ZappoMan** [comment]
@howard-stearns that appears to be in the Connexion code... Isn't that the 3d mouse?

**Comment by jherico** [comment]
That may be unrelated, I merged some 3D mouse code last night.  should have done more testing I guess.

**Comment by ZappoMan** [comment]
@howard-stearns - the repro that matters is doing a "stop" or "restart" in a script in the running script dialog.

**Comment by howard-stearns** [comment]
I imagine it is unrelated. I'm reporting it because it means that I cannot test "shutdown in interface" because of it, which is one of the three things this is meant to fix. Maybe we can address the connexion code and merge that here, so that we can properly test this?

**Comment by howard-stearns** [comment]
Oh, I see we've merged anyway. I guess we'll just wing it... ok.

**Comment by ZappoMan** [comment]
@howard-stearns - I think this is a build system problem... we need @murillodigital to look at it.. looks like the package doesn't include the dynlib.

**Pull request merged**